clippy_lints/src/manual_unwrap_or_default.rs

```diff
@@ -131,6 +131,17 @@ fn handle<'tcx>(cx: &LateContext<'tcx>, if_let_or_match: IfLetOrMatch<'tcx>, exp
         },
     };
 
+    // We check if the expression is not a None variant
+    if let Some(none_def_id) = cx.tcx.lang_items().option_none_variant() {
+        if let ExprKind::Path(QPath::Resolved(_, path)) = &condition.kind {
+            if let Some(def_id) = path.res.opt_def_id() {
+                if cx.tcx.parent(def_id) == none_def_id {
+                    return;
```

Comment:
What about just changing the applicability away from `MachineApplicable`, instead of the `return` and not linting at all? I think it is still useful to emit a warning, the user would just need to provide type annotations, e.g. write `None::<i32>.unwrap_or_default()`.

Reply:
Oh sorry, I didn't mean that it needed to produce a machine applicable suggestion for that specific `None` case.

I can't imagine a situation where someone would actually want to write `if let Some(...) = None`, so I don't think it would really be worth adding much logic for that specific case. In fact, ideally we shouldn't need to special-case `None` at all, see my second comment.

I only meant using a *different* applicability for these cases, like `Applicability::HasPlaceholder`, similar to how it conditionally chooses an applicability for when there is a comment:

https://github.com/rust-lang/rust-clippy/blob/cdd6336386a0cdd164f912bab1038e02871e1ee4/clippy_lints/src/manual_unwrap_or_default.rs#L151-L156

That way, by using some other applicability, `--fix` won't try to auto-apply the suggestion and the user will have to manually apply the suggestion themselves (*e.g.* by writing `None::<i32>` - but this lint doesnt' need to bother suggesting it, because in practice it will most likely be some other kind of expression)

(It would also need to be in its own test file, with `//@no-rustfix` at the top, similar to this test: https://github.com/rust-lang/rust-clippy/blob/master/tests/ui/unnecessary_cast_unfixable.rs#L2)

Reply:
I put the HasPlaceholder for the None, but isn't MachineApplicable better for the `.parse()` case you provided ?